File: src-utools/src/dsh.js
/**
 * dsh 生态访问层：DSH_HOME 解析、profile 清单读写、cordis patch 读写、
 * bundle 行 id 解析、已安装包信息、dsh 可执行文件解析与校验。
 * 所有函数在 preload（Node）环境执行，页面侧经 window.preload.dsh 调用。
 */
const { homedir, platform } = require('node:os')
const { join } = require('node:path')
const {
  existsSync,
  readFileSync,
  writeFileSync,
  readdirSync,
  accessSync,
  constants
} = require('node:fs')
const { spawnSync } = require('node:child_process')
const { userEnv } = require('./process')

const DSH_HOME_DIR_NAME = '.dsh'
const PROFILE_PATCH_FILENAME = 'cordis.patch.yml'

/** $DSH_HOME：显式配置 > $DSH_HOME 环境变量 > ~/.dsh */
function resolveDshHome() {
  const fromEnv = process.env.DSH_HOME
  if (fromEnv && fromEnv.trim()) return fromEnv.trim()
  return join(homedir(), DSH_HOME_DIR_NAME)
}

function profilesDir() {
  return join(resolveDshHome(), 'profiles')
}

/** 列出所有 profile（目录下存在 package.json 视为一个 profile） */
function listProfiles() {
  const dir = profilesDir()
  if (!existsSync(dir)) return []
  return readdirSync(dir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory() && existsSync(join(dir, entry.name, 'package.json')))
    .map((entry) => entry.name)
    .sort()
}

function profileDir(profile) {
  return join(profilesDir(), profile)
}

function readProfileManifest(profile) {
  const file = join(profileDir(profile), 'package.json')
  if (!existsSync(file)) return null
  try {
    return JSON.parse(readFileSync(file, 'utf8'))
  } catch {
    return null
  }
}

function writeProfileManifest(profile, manifest) {
  const file = join(profileDir(profile), 'package.json')
  writeFileSync(file, JSON.stringify(manifest, null, 2) + '\n')
}

function profilePatchPath(profile) {
  return join(profileDir(profile), PROFILE_PATCH_FILENAME)
}

/** 读取 profile 级 cordis.patch.yml 原文；不存在时返回空串 */
function readProfilePatch(profile) {
  const file = profilePatchPath(profile)
  if (!existsSync(file)) return ''
  return readFileSync(file, 'utf8')
}

function writeProfilePatch(profile, content) {
  const file = profilePatchPath(profile)
  writeFileSync(file, content)
}

/**
 * 解析某个 bundle 在 profile 中的安装目录。
 * pnpm 的 hoisted 结构下，@deepseek-ai/* 位于 profiles/node_modules，
 * 第三方包位于 profiles/<profile>/node_modules，两处兜底。
 */
function resolveBundleDir(profile, packageName) {
  const candidates = [
    join(profileDir(profile), 'node_modules', packageName),
    join(profilesDir(), 'node_modules', packageName)
  ]
  for (const candidate of candidates) {
    if (existsSync(join(candidate, 'package.json'))) return candidate
  }
  return null
}

/** 读取已安装包（bundle 或普通依赖）的 package.json */
function readInstalledPackage(profile, packageName) {
  const dir = resolveBundleDir(profile, packageName)
  if (!dir) return null
  try {
    return JSON.parse(readFileSync(join(dir, 'package.json'), 'utf8'))
  } catch {
    return null
  }
}

/** 读取 bundle 自带 patch（dsh.bundle.patch 指向的文件）原文 */
function readBundlePatch(profile, packageName) {
  const dir = resolveBundleDir(profile, packageName)
  if (!dir) return null
  let manifest
  try {
    manifest = JSON.parse(readFileSync(join(dir, 'package.json'), 'utf8'))
  } catch {
    return null
  }
  const patchRel = manifest.dsh?.bundle?.patch
  if (!patchRel) return null
  const patchFile = join(dir, patchRel)
  if (!existsSync(patchFile)) return null
  return readFileSync(patchFile, 'utf8')
}

function isExecutable(file) {
  try {
    accessSync(file, constants.X_OK)
    return true
  } catch {
    return false
  }
}

/** 解析 dsh 可执行文件：手动配置 > PATH > ~/.bun/bin（bun 全局安装目录） */
function resolveDshBin(configuredPath) {
  if (configuredPath) {
    if (existsSync(configuredPath) && isExecutable(configuredPath)) return configuredPath
    return null
  }
  const fromPath = findOnPath('dsh')
  if (fromPath) return fromPath
  const bunBin = join(homedir(), '.bun', 'bin', platform() === 'win32' ? 'dsh.exe' : 'dsh')
  if (existsSync(bunBin) && isExecutable(bunBin)) return bunBin
  return null
}

/** 在 PATH 上查找可执行文件（通用） */
function findOnPath(name) {
  const pathVar = process.env.PATH || ''
  const delimiter = platform() === 'win32' ? ';' : ':'
  for (const dir of pathVar.split(delimiter)) {
    if (!dir) continue
    for (const candidate of platform() === 'win32'
      ? [`${name}.exe`, `${name}.cmd`, `${name}.bat`]
      : [name]) {
      const file = join(dir, candidate)
      if (existsSync(file) && isExecutable(file)) return file
    }
  }
  return null
}

/** 查找 bun 可执行文件（bun 安装的 dsh 无 node 时用它兜底解释执行） */
function findBun() {
  const bunBin = join(homedir(), '.bun', 'bin', platform() === 'win32' ? 'bun.exe' : 'bun')
  if (existsSync(bunBin) && isExecutable(bunBin)) return bunBin
  return findOnPath('bun')
}

/**
 * 探测 dsh 的实际运行方式：
 * 1. 直接执行（注入用户 PATH，shebang 的 env node 可解析）；
 * 2. 失败（常见于只装 bun 未装 node）→ 用 bun 解释执行。
 * @returns {{ok: boolean, command?: string, prefix?: string[], version?: string, error?: string}}
 */
function probeDsh(binPath) {
  const opts = { encoding: 'utf8', timeout: 10000, env: userEnv() }
  const direct = spawnSync(binPath, ['--version'], opts)
  if (!direct.error && direct.status === 0) {
    return {
      ok: true,
      command: binPath,
      prefix: [],
      version: (direct.stdout || '').trim().split('\n')[0] || 'unknown'
    }
  }
  // 直接执行失败：尝试 bun 解释执行（bun 兼容 node 脚本）
  const bun = findBun()
  if (bun) {
    const viaBun = spawnSync(bun, [binPath, '--version'], opts)
    if (!viaBun.error && viaBun.status === 0) {
      return {
        ok: true,
        command: bun,
        prefix: [binPath],
        version: (viaBun.stdout || '').trim().split('\n')[0] || 'unknown'
      }
    }
  }
  const reason = (direct.stderr || direct.error?.message || '').trim().slice(0, 200)
  return { ok: false, error: reason || `exit code ${direct.status}` }
}

/**
 * 解析并探测 dsh：手动配置 > PATH > ~/.bun/bin，随后探测运行方式。
 * @returns {{state: 'ok'|'missing'|'invalid', path?: string, command?: string, prefix?: string[], version?: string, error?: string}}
 */
function resolveDsh(configuredPath) {
  const binPath = resolveDshBin(configuredPath)
  if (!binPath) {
    return configuredPath
      ? { state: 'invalid', path: configuredPath, error: 'file-not-found' }
      : { state: 'missing' }
  }
  const probed = probeDsh(binPath)
  if (!probed.ok) {
    return { state: 'invalid', path: binPath, error: probed.error }
  }
  return {
    state: 'ok',
    path: binPath,
    command: probed.command,
    prefix: probed.prefix,
    version: probed.version
  }
}

module.exports = {
  getDshHome: resolveDshHome,
  listProfiles,
  readProfileManifest,
  writeProfileManifest,
  readProfilePatch,
  writeProfilePatch,
  readBundlePatch,
  readInstalledPackage,
  resolveDshBin,
  resolveDsh
}

File: services/dsh.go
/**
 * dsh 生态访问服务：DSH_HOME 解析、profile 清单读写、cordis patch 读写、
 * bundle 行 id 解析、已安装包信息、dsh 可执行文件解析与校验。
 * 本文件为原 src-utools/src/dsh.js 的 Go 翻译，行为保持一致。
 */
package services

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	dshHomeDirName       = ".dsh"
	profilePatchFilename = "cordis.patch.yml"
)

// DshResolveResult dsh 可执行文件解析结果（state: ok | missing | invalid）。
type DshResolveResult struct {
	State   string   `json:"state"`
	Path    string   `json:"path,omitempty"`
	Command string   `json:"command,omitempty"`
	Prefix  []string `json:"prefix,omitempty"`
	Version string   `json:"version,omitempty"`
	Error   string   `json:"error,omitempty"`
}

type probeResult struct {
	ok      bool
	command string
	prefix  []string
	version string
	errMsg  string
}

// DshService dsh 生态访问层。
type DshService struct{}

func NewDshService() *DshService {
	return &DshService{}
}

/** $DSH_HOME：显式配置 > $DSH_HOME 环境变量 > ~/.dsh */
func (s *DshService) GetDshHome() string {
	if fromEnv := strings.TrimSpace(os.Getenv("DSH_HOME")); fromEnv != "" {
		return fromEnv
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return dshHomeDirName
	}
	return filepath.Join(home, dshHomeDirName)
}

func (s *DshService) profilesDir() string {
	return filepath.Join(s.GetDshHome(), "profiles")
}

func (s *DshService) profileDir(profile string) string {
	return filepath.Join(s.profilesDir(), profile)
}

/** 列出所有 profile（目录下存在 package.json 视为一个 profile）。 */
func (s *DshService) ListProfiles() []string {
	dir := s.profilesDir()
	entries, err := os.ReadDir(dir)
	if err != nil {
		return []string{}
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		if _, err := os.Stat(filepath.Join(dir, e.Name(), "package.json")); err == nil {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	return names
}

/** 读取 profile 的 package.json；不存在或解析失败返回 nil。 */
func (s *DshService) ReadProfileManifest(profile string) map[string]any {
	raw, err := os.ReadFile(filepath.Join(s.profileDir(profile), "package.json"))
	if err != nil {
		return nil
	}
	var manifest map[string]any
	if err := json.Unmarshal(raw, &manifest); err != nil {
		return nil
	}
	return manifest
}

func (s *DshService) WriteProfileManifest(profile string, manifest map[string]any) error {
	raw, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(s.profileDir(profile), "package.json"), append(raw, '\n'), 0o644)
}

/** 读取 profile 级 cordis.patch.yml 原文；不存在时返回空串。 */
func (s *DshService) ReadProfilePatch(profile string) string {
	raw, err := os.ReadFile(filepath.Join(s.profileDir(profile), profilePatchFilename))
	if err != nil {
		return ""
	}
	return string(raw)
}

func (s *DshService) WriteProfilePatch(profile string, content string) error {
	return os.WriteFile(filepath.Join(s.profileDir(profile), profilePatchFilename), []byte(content), 0o644)
}

/**
 * 解析某个 bundle 在 profile 中的安装目录。
 * pnpm 的 hoisted 结构下，@deepseek-ai/* 位于 profiles/node_modules，
 * 第三方包位于 profiles/<profile>/node_modules，两处兜底。
 */
func (s *DshService) resolveBundleDir(profile, packageName string) string {
	candidates := []string{
		filepath.Join(s.profileDir(profile), "node_modules", packageName),
		filepath.Join(s.profilesDir(), "node_modules", packageName),
	}
	for _, c := range candidates {
		if _, err := os.Stat(filepath.Join(c, "package.json")); err == nil {
			return c
		}
	}
	return ""
}

/** 读取已安装包（bundle 或普通依赖）的 package.json；未安装返回 nil。 */
func (s *DshService) ReadInstalledPackage(profile, packageName string) map[string]any {
	dir := s.resolveBundleDir(profile, packageName)
	if dir == "" {
		return nil
	}
	raw, err := os.ReadFile(filepath.Join(dir, "package.json"))
	if err != nil {
		return nil
	}
	var pkg map[string]any
	if err := json.Unmarshal(raw, &pkg); err != nil {
		return nil
	}
	return pkg
}

/** 读取 bundle 自带 patch（dsh.bundle.patch 指向的文件）原文；无则返回 nil。 */
func (s *DshService) ReadBundlePatch(profile, packageName string) *string {
	dir := s.resolveBundleDir(profile, packageName)
	if dir == "" {
		return nil
	}
	raw, err := os.ReadFile(filepath.Join(dir, "package.json"))
	if err != nil {
		return nil
	}
	var manifest map[string]any
	if err := json.Unmarshal(raw, &manifest); err != nil {
		return nil
	}
	dsh, _ := manifest["dsh"].(map[string]any)
	if dsh == nil {
		return nil
	}
	bundle, _ := dsh["bundle"].(map[string]any)
	if bundle == nil {
		return nil
	}
	patchRel, _ := bundle["patch"].(string)
	if patchRel == "" {
		return nil
	}
	content, err := os.ReadFile(filepath.Join(dir, patchRel))
	if err != nil {
		return nil
	}
	str := string(content)
	return &str
}

func isExecutable(path string) bool {
	info, err := os.Stat(path)
	if err != nil || info.IsDir() {
		return false
	}
	if runtime.GOOS == "windows" {
		return true
	}
	return info.Mode()&0o111 != 0
}

func dshName() string {
	if runtime.GOOS == "windows" {
		return "dsh.exe"
	}
	return "dsh"
}

func bunName() string {
	if runtime.GOOS == "windows" {
		return "bun.exe"
	}
	return "bun"
}

/** 在 PATH 上查找可执行文件（通用）。 */
func (s *DshService) findOnPath(name string) string {
	delim := ":"
	if runtime.GOOS == "windows" {
		delim = ";"
	}
	for _, dir := range strings.Split(os.Getenv("PATH"), delim) {
		if dir == "" {
			continue
		}
		candidates := []string{name}
		if runtime.GOOS == "windows" {
			candidates = []string{name + ".exe", name + ".cmd", name + ".bat"}
		}
		for _, cand := range candidates {
			full := filepath.Join(dir, cand)
			if isExecutable(full) {
				return full
			}
		}
	}
	return ""
}

/** 查找 bun 可执行文件（bun 安装的 dsh 无 node 时用它兜底解释执行）。 */
func (s *DshService) findBun() string {
	home, _ := os.UserHomeDir()
	bunBin := filepath.Join(home, ".bun", "bin", bunName())
	if isExecutable(bunBin) {
		return bunBin
	}
	return s.findOnPath("bun")
}

/** 解析 dsh 可执行文件：手动配置 > PATH > ~/.bun/bin（bun 全局安装目录）。 */
func (s *DshService) ResolveDshBin(configuredPath string) *string {
	if configuredPath != "" {
		if isExecutable(configuredPath) {
			return &configuredPath
		}
		return nil
	}
	if fromPath := s.findOnPath("dsh"); fromPath != "" {
		return &fromPath
	}
	home, _ := os.UserHomeDir()
	bunBin := filepath.Join(home, ".bun", "bin", dshName())
	if isExecutable(bunBin) {
		return &bunBin
	}
	return nil
}

/** 执行 `<bin> [prefix...] --version` 并返回首行版本号（10s 超时，注入用户 PATH）。 */
func runVersion(bin string, prefix []string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	args := append(append([]string{}, prefix...), "--version")
	c := exec.CommandContext(ctx, bin, args...)
	c.Env = userEnv(nil)
	out, err := c.CombinedOutput()
	if err != nil {
		return "", err
	}
	if c.ProcessState.ExitCode() != 0 {
		return "", fmt.Errorf("exit code %d", c.ProcessState.ExitCode())
	}
	version := strings.TrimSpace(strings.Split(string(out), "\n")[0])
	if version == "" {
		version = "unknown"
	}
	return version, nil
}

func describeVersionErr(err error) string {
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		if stderr := strings.TrimSpace(string(ee.Stderr)); stderr != "" {
			return stderr[:min(len(stderr), 200)]
		}
		return fmt.Sprintf("exit code %d", ee.ExitCode())
	}
	return err.Error()
}

/**
 * 探测 dsh 的实际运行方式：
 * 1. 直接执行（注入用户 PATH，shebang 的 env node 可解析）；
 * 2. 失败（常见于只装 bun 未装 node）→ 用 bun 解释执行。
 */
func (s *DshService) probeDsh(binPath string) probeResult {
	if version, err := runVersion(binPath, nil); err == nil {
		return probeResult{ok: true, command: binPath, prefix: []string{}, version: version}
	} else {
		if bun := s.findBun(); bun != "" {
			if version, err := runVersion(bun, []string{binPath}); err == nil {
				return probeResult{ok: true, command: bun, prefix: []string{binPath}, version: version}
			}
		}
		return probeResult{ok: false, errMsg: describeVersionErr(err)}
	}
}

/** 解析并探测 dsh：手动配置 > PATH > ~/.bun/bin，随后探测运行方式。 */
func (s *DshService) ResolveDsh(configuredPath string) DshResolveResult {
	binPath := s.ResolveDshBin(configuredPath)
	if binPath == nil {
		if configuredPath != "" {
			return DshResolveResult{State: "invalid", Path: configuredPath, Error: "file-not-found"}
		}
		return DshResolveResult{State: "missing"}
	}
	probed := s.probeDsh(*binPath)
	if !probed.ok {
		return DshResolveResult{State: "invalid", Path: *binPath, Error: probed.errMsg}
	}
	return DshResolveResult{
		State:   "ok",
		Path:    *binPath,
		Command: probed.command,
		Prefix:  probed.prefix,
		Version: probed.version,
	}
}
